consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,209 @@
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, epochNumber uint64) error {
+	checkpointID, err := c.getCurrentCheckpointID(epochNumber)
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	// detect any pending (previously failed) checkpoints and send them
+	for blockNum := checkpointID + 1; blockNum < latestHeader.Number; blockNum++ {
+		header, exists := c.blockchain.GetHeaderByNumber(blockNum)
+		if !exists {
+			return fmt.Errorf("block %d was not found", blockNum)
+		}
+
+		extra, err := GetIbftExtra(header.ExtraData)
+		if err != nil {
+			return err
+		}
+
+		// send checkpoints only for epoch ending blocks (validators delta is non-empty)
+		if extra.Validators.IsEmpty() {
+			continue
+		}
+
+		err = c.submitCheckpointInternal(nonce, txn, *header, *extra)
+		if err != nil {
+			return err
+		}
+		nonce++
+	}
+
+	extra, err := GetIbftExtra(latestHeader.ExtraData)
+	if err != nil {
+		return err
+	}
+
+	return c.submitCheckpointInternal(nonce, txn, latestHeader, *extra)
+}
+
+// submitCheckpointInternal encodes checkpoint data for the given block and
+// sends a transaction to the CheckpointManager rootchain contract
+func (c *checkpointManager) submitCheckpointInternal(nonce uint64, txn *ethgo.Transaction,
+	header types.Header, extra Extra) error {
+	nextEpochValidators, err := c.consensusBackend.GetValidators(header.Number, nil)
+	if err != nil {
+		return err
+	}
+
+	input, err := c.abiEncodeCheckpointBlock(header, extra, nextEpochValidators)
+	if err != nil {
+		return fmt.Errorf("failed to encode checkpoint data to ABI for block %d: %w", header.Number, err)
+	}
+
+	txn.Input = input
+
+	receipt, err := c.rootchain.SendTransaction(nonce, txn)
+	if err != nil {
+		return err
+	}
+
+	if receipt.Status == uint64(types.ReceiptFailed) {
+		return fmt.Errorf("transaction execution failed for block %d", header.Number)
+	}
+
+	return nil
+}
+
+// abiEncodeCheckpointBlock encodes checkpoint data into ABI format for a given header
+func (c *checkpointManager) abiEncodeCheckpointBlock(header types.Header, extra Extra,
+	nextValidators AccountSet) ([]byte, error) {
+	params := map[string]interface{}{
+		"chainID":             new(big.Int).SetUint64(c.blockchain.GetChainID()),
+		"aggregatedSignature": extra.Committed.AggregatedSignature,
+		"validatorsBitmap":    extra.Committed.Bitmap,
+		"epochNumber":         new(big.Int).SetUint64(extra.Checkpoint.EpochNumber),
+		"blockNumber":         new(big.Int).SetUint64(header.Number),
+		"blockHash":           header.Hash,
+		"blockRound":          new(big.Int).SetUint64(extra.Checkpoint.BlockRound),
+		"eventRoot":           extra.Checkpoint.EventRoot.Bytes(),
```

Comment:
Is it bytes or bytes32 in solidity?

Reply:
`eventRoot` is `bytes32` solidity type (https://github.com/0xPolygon/polygon-edge/pull/861/files#diff-fce12e7457acbf9d31265d9e0f71a4340b87f09cb3dd52e2f61df69d0bbc142dR25), but apparently `ethgo` knows how to deal with it internally.